### PR TITLE
[EPG] Fix deadlock epg mutex vs channelgroup container mutex.

### PR DESCRIPTION
**Pull request by ksooo**
http://pastebin.com/3YEPCN2j => deadlock thread 1 vs. thread 24

@fritsch could you please test?

### CODE REVIEW

**Comment by fritsch** [comment]
@ksooo sadly I could not reproduce the original issue as well
@MilhouseVH could you include this in a testbuild? I will send the user with the issue to test this commit.

**Comment by MilhouseVH** [comment]
@fritsch: yes, but won't be until tomorrow, possibly overnight. Presumably a Generic build?

File: xbmc/epg/Epg.cpp [hunk 1]
```diff
@@ -331,36 +331,21 @@ bool CEpg::UpdateEntry(const EPG_TAG *data, bool bUpdateDatabase /* = false */)
...
-    }
-    return true;
-  }
-  return false;
+  return UpdateEntry(tag, it == m_tags.end() ? EPG_EVENT_CREATED : EPG_EVENT_UPDATED, it, bUpdateDatabase);
 }
 
 bool CEpg::UpdateEntry(const CEpgInfoTagPtr &tag, EPG_EVENT_STATE newState, bool bUpdateDatabase /* = false */)
 {
-  CSingleLock lock(m_critSection);
```
Comment (tamland) [inline xbmc/epg/Epg.cpp]: that doesn't look very safe. you're access m_tags without a lock now

**Comment by mule1970** [comment]
Just tried the newest milhouse build with this patch. Sorry to say but channel loading times are slow as within the previous build with async process and kodi is crashing once again when trying to call the tvmanager while channels are loading.

Crashlog: [url]http://pastebin.com/tuDhutHa[/url]

Anything else i can/should do/try?

**Comment by FernetMenta** [comment]
the problem is incorrect usage of smart pointers as done here. The smart pointer needs to be assigned to a local variable before leaving the lock.

```
CEpgPtr CEpgContainer::GetById(int iEpgId) const
{
  if (iEpgId < 0)
    return NULL;

  CSingleLock lock(m_critSection);
  const auto &epgEntry = m_epgs.find((unsigned int) iEpgId);
  return epgEntry != m_epgs.end() ? epgEntry->second : NULL;
}
```

**Comment by ksooo** [comment]
> Sorry to say but channel loading times are slow as within the previous build with async process

 nobody said that this patch would speed up things. ;-)

**Comment by ksooo** [comment]
Seems that my assumption that this is a deadlock was wrong.

**Comment by ksooo** [comment]
@FernetMenta 

> The smart pointer needs to be assigned to a local variable before leaving the lock.

 For the example you gave, does this mean that the mutex is released first, and then <code>m_epgs.end()</code> and <code>epgEntry->second</code> are executed without mutex protection?

**Comment by FernetMenta** [comment]
Nope, that means that the lock is left before the smart pointer increases refCount. So the object may not exist anymore when finally being assigned to another smart pointer.

**Comment by ksooo** [comment]
That's what I thought, but sorry, don't get it, then. Could you please explain a little more in detail why the example you gave is problematic? Cannot see where a refcount is increased after lock is destructed.

**Comment by FernetMenta** [comment]
Assigning a smart pointer increases ref count. When the last reference is removed, the object gets destructed, right?

Now you do this. With assignment to temp, the ref count is increased and the object is safe. 
CEpgPtr tmp = container.GetById(id);

The problem is that you already left the lock before the assignment is completed. There is more of those, i.e.:

```
CEpgInfoTagPtr CPVRGUIInfo::GetPlayingTag() const
{
  CSingleLock lock(m_critSection);
  return m_playingEpgTag;
}
```

This lock does not help anything either.

**Comment by ksooo** [comment]
Just stepped through GetPlayingTag and disassembly looks like assignment and increment refcount is called before dtor of lock.

<pre>
Kodi`PVR::CPVRGUIInfo::GetPlayingTag:
    0x1016265a0 <+0>:   pushq  %rbp
    0x1016265a1 <+1>:   movq   %rsp, %rbp
    0x1016265a4 <+4>:   subq   $0x60, %rsp
    0x1016265a8 <+8>:   movq   %rdi, %rax
    0x1016265ab <+11>:  leaq   -0x38(%rbp), %rcx
    0x1016265af <+15>:  movq   %rsi, -0x28(%rbp)
    0x1016265b3 <+19>:  movq   -0x28(%rbp), %rsi
    0x1016265b7 <+23>:  movq   %rsi, %rdx
    0x1016265ba <+26>:  addq   $0x1a40, %rdx             ; imm = 0x1A40 
    0x1016265c1 <+33>:  movq   %rdi, -0x40(%rbp)
    0x1016265c5 <+37>:  movq   %rcx, %rdi
    0x1016265c8 <+40>:  movq   %rsi, -0x48(%rbp)
    0x1016265cc <+44>:  movq   %rdx, %rsi
    0x1016265cf <+47>:  movq   %rax, -0x50(%rbp)
    0x1016265d3 <+51>:  callq  0x100043350               ; CSingleLock::CSingleLock at SingleLock.h:39
->  0x1016265d8 <+56>:  movq   -0x48(%rbp), %rax
    0x1016265dc <+60>:  addq   $0x19b0, %rax             ; imm = 0x19B0 
    0x1016265e2 <+66>:  movq   -0x40(%rbp), %rcx
    0x1016265e6 <+70>:  movq   %rcx, -0x18(%rbp)
    0x1016265ea <+74>:  movq   %rax, -0x20(%rbp)
    0x1016265ee <+78>:  movq   -0x18(%rbp), %rax
    0x1016265f2 <+82>:  movq   -0x20(%rbp), %rdx
    0x1016265f6 <+86>:  movq   %rax, -0x8(%rbp)
    0x1016265fa <+90>:  movq   %rdx, -0x10(%rbp)
    0x1016265fe <+94>:  movq   -0x8(%rbp), %rax
    0x101626602 <+98>:  movq   -0x10(%rbp), %rdx
    0x101626606 <+102>: movq   (%rdx), %rdx
    0x101626609 <+105>: movq   %rdx, (%rax)
    0x10162660c <+108>: movq   -0x10(%rbp), %rdx
    0x101626610 <+112>: movq   0x8(%rdx), %rdx
    0x101626614 <+116>: movq   %rdx, 0x8(%rax)
    0x101626618 <+120>: cmpq   $0x0, 0x8(%rax)
    0x10162661d <+125>: movq   %rax, -0x58(%rbp)
    0x101626621 <+129>: je     0x101626634               ; <+148> at memory:4254
    0x101626627 <+135>: movq   -0x58(%rbp), %rax
    0x10162662b <+139>: movq   0x8(%rax), %rdi
    0x10162662f <+143>: callq  0x103742cd0               ; symbol stub for: std::__1::__shared_weak_count::__add_shared()
    0x101626634 <+148>: leaq   -0x38(%rbp), %rdi
    0x101626638 <+152>: callq  0x100040f80               ; CSingleLock::~CSingleLock at SingleLock.h:35
    0x10162663d <+157>: movq   -0x50(%rbp), %rax
    0x101626641 <+161>: addq   $0x60, %rsp
    0x101626645 <+165>: popq   %rbp
    0x101626646 <+166>: retq 
</pre>

Or am I still missing the point?

**Comment by ksooo** [comment]
@FernetMenta okay, stupid me. Got it now, I guess. For some stupid reason I was assuming that the compiler creates a copy of the smart pointer and returns the copy. :-/

So, a correct implementation of GetPlayingTag, would be

<pre>
CEpgInfoTagPtr CPVRGUIInfo::GetPlayingTag() const
{
  CSingleLock lock(m_critSection);
  return CEpgInfoTagPtr(m_playingEpgTag);
}
</pre>

Right?

We have many places in our code base where this is not handled correctly, I'm afraid...

**Comment by FernetMenta** [comment]
@ksooo yes, with RVO we can't rely on temporary objects anymore. The problem with your suggestion above is, that it creates additional overhead by increment/decrement the internal refcount.

**Comment by ksooo** [comment]
So, how to do it "right"?

**Comment by ksooo** [comment]
Not returning smart pointer but pass non const ref to smart pointer as method arguments?

**Comment by FernetMenta** [comment]
>  pass non const ref to smart pointer as method arguments

that's imo the best option for now

**Comment by ksooo** [comment]
@MaxKellermann do you have an idea for a more elegant solution than to change all method signatures not to return smart pointer instances but to use out parameters to achieve thread safety and not to add additional overhead?

**Comment by MaxKellermann** [comment]
If you pass a const reference to a smart pointer, who's responsible for protecting that smart pointer? Remember: a smart pointer is not thread-safe - only its control block is!

**Comment by ksooo** [comment]
@MaxKellermann the idea is to change for example

<code>CEpgInfoTagPtr GetFoo() const</code> to <code>void GetFoo(CEpgInfoTagPtr &retVal) const</code>.

Original implementation looks like

<pre>
CEpgInfoTagPtr Someclass::GetFoo() const
{
  CCriticalSecition lock(m_mutex);
  return m_someMember;
}
</pre>

Ofc you know that if you call GetFoo like this

<pre>
...
CEpgTagInfoPtr tag = GetFoo();
...
</pre>

and recount of m_somemember is 1 (and RVO kicks in, thus no copy ctor is called an return, thus refcount stays 1), first the mutex gets released, then assignement operator is called to increment refcount. If another thread resets m_somemember after mutex got released and before assignemt operator is called the heap object gets destroyed and after this assignment with the destructed heap object is tried.

New implementation could look like

<pre>
void Someclass::GetFoo(CEpgInfoTagPtr &retVal) const
{
  CCriticalSecition lock(m_mutex);
  retVal = m_someMember;
}
</pre>

Idea is that retVal gets assigned a value while m_mutex is locked to achieve thread safety.

Makes sense?

**Comment by MaxKellermann** [comment]
Hm, I don't understand what problem you're trying to solve. And I don't understand why RVO poses a problem here. Are you discussing a theoretical problem, or did you already observe that a compiler unlocks the mutex before copying the `shared_ptr`?

**Comment by MaxKellermann** [comment]
You may be misunderstanding the C++ standard, and a clarification of the wording is already being discussed to avoid this misunderstanding: http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1885
Is this what you're worried about?

**Comment by fritsch** [comment]
To get back on topic, basically a solution for: http://pastebin.com/3YEPCN2j is searched, which is not a deadlock, but a smart_ptr not refcounted enough ;-) going out of scope and removing the objects under the callers ass.

**Comment by FernetMenta** [comment]
> The C++ standard allows an implementation to omit creating a temporary which is only used to initialize another object of the same type.


see -fno-elide-constructors

**Comment by MaxKellermann** [comment]
> see -fno-elide-constructors
Why would you want to disable this optimization?

**Comment by fritsch** [comment]
>GCC provides the -fno-elide-constructors option to disable copy-elision. This option is useful to observe (or not observe!) the effects of Return Value Optimization or other optimizations where copies are elided. It is generally not recommended to disable this important optimization.

**Comment by FernetMenta** [comment]
I don't want to disable this. I am just saying that this is the root cause of this issue here. The internal ref count of the shared pointer is not increased before the lock is left.

**Comment by MaxKellermann** [comment]
> The internal ref count of the shared pointer is not increased before the lock is left. 

What makes you sure of this? Did you confirm this in the disassembly?

**Comment by fritsch** [comment]
I checked the backtrace and interpreted the code. It's my interpretation. Do you see any other possibility why a mutex might be "gone" out of nowhere?

**Comment by MaxKellermann** [comment]
If that's your interpretation, then read the disassembly to verify your interpretation.

**Comment by fritsch** [comment]
I would love to do this - but sadly I cannot reproduce the issue so need to
take what was given: the backtrace only.

2016-04-12 15:18 GMT+02:00 Max Kellermann <notifications@github.com>:

> If that's your interpretation, then read the disassembly to verify your
> interpretation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/9589#issuecomment-208900757>
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by FernetMenta** [comment]
> When the function terminates, the following steps happen:
Registers are restored from the call stack
The stack frame is popped off the stack. This destroys all local variables and arguments.
The return value is handled.
The CPU resumes execution at the return address.

source: http://www.learncpp.com/cpp-tutorial/79-the-stack-and-the-heap/

local variable are destructed before the return value is handled.

The fact that this is in doubt is reason enough to change it.

**Comment by MaxKellermann** [comment]
@FernetMenta your source is wrong. There is no doubt.

**Comment by MaxKellermann** [comment]
@fritsch then let me disassemble for you. I downloaded http://fritsch.fruehberger.net/openelec/v17_videoplayer_testing/OpenELEC-Generic.x86_64-7.0-devel-20160403092304-r22336-g1112080.tar (the one which generated the backtrace), extracted kodi.bin and disassembled the function `CPVRGUIInfo::GetPlayingTag()`, as an example:

    b49a30:       55                      push   %rbp
    b49a31:       53                      push   %rbx
    b49a32:       48 89 fd                mov    %rdi,%rbp
    b49a35:       48 8d be 20 18 00 00    lea    0x1820(%rsi),%rdi
    b49a3c:       48 89 f3                mov    %rsi,%rbx
    b49a3f:       48 83 ec 18             sub    $0x18,%rsp
    b49a43:       48 89 3c 24             mov    %rdi,(%rsp)
    b49a47:       c6 44 24 08 01          movb   $0x1,0x8(%rsp)
    b49a4c:       e8 77 36 cc ff          callq  80d0c8 <_ZN11XbmcThreads16CountingLockableINS_8pthreads14RecursiveMutexEE4lockEv@@Base>
    b49a51:       48 8b 83 c0 17 00 00    mov    0x17c0(%rbx),%rax
    b49a58:       48 8d b3 c8 17 00 00    lea    0x17c8(%rbx),%rsi
    b49a5f:       48 8d 7d 08             lea    0x8(%rbp),%rdi
    b49a63:       48 89 45 00             mov    %rax,0x0(%rbp)
    b49a67:       e8 7c 27 cc ff          callq  80c1e8 <_ZNSt14__shared_countILN9__gnu_cxx12_Lock_policyE2EEC1ERKS2_@@Base>
    b49a6c:       48 89 e7                mov    %rsp,%rdi
    b49a6f:       e8 6e 35 cc ff          callq  80cfe2 <_ZN11XbmcThreads10UniqueLockI16CCriticalSectionED1Ev@@Base>
    b49a74:       48 83 c4 18             add    $0x18,%rsp
    b49a78:       48 89 e8                mov    %rbp,%rax
    b49a7b:       5b                      pop    %rbx
    b49a7c:       5d                      pop    %rbp
    b49a7d:       c3                      retq   

Is that proof enough that your "return/destructor" theory is wrong?

**Comment by fritsch** [comment]
Thanks. Then we need a new theory now. As the bug is not yet solved.

**Comment by MaxKellermann** [comment]
My theory is: use-after-free. (Just guessing, but the pthread_mutex_t is corrupt, and many corruptions are really use-after-free.)

**Comment by MaxKellermann** [comment]
The backtrace indeed looks like deadlock at first glance, but I don't think it's a deadlock. The code has potential for deadlocks, and must be fixed, but a deadlock is not what we observe here. It would be deadlocking if both threads would be accessing the same `CEpg` instance. What we see is a corrupt `pthread_mutex_t`. That's a different story.

**Comment by FernetMenta** [comment]
@fritsch did you compile the version that crashed? otherwise the disassembly is not a proof.

**Comment by fritsch** [comment]
Yes, that's the version I compiled.

**Comment by ksooo** [comment]
So, if nobody comes up with a more elegant (wrt how "natural" the code to write feels) and at least as efficient approach, I will solve the problem using the "non-const ref param" approach. Okay @FernetMenta?

**Comment by fritsch** [comment]
I would like to understand the real problem first before shooting with canons. I am curious on how @MaxKellermann would approach solving this.

**Comment by MaxKellermann** [comment]
@ksooo your "non-const ref param" approach is wrong because the problem it addresses doesn't exist. There is no return/destructor ordering problem!

**Comment by ksooo** [comment]
Forget my last comment. Did not see that the discussion had moved forward already.

**Comment by MaxKellermann** [comment]
@fritsch The problem here is that the crash log doesn't give a clue where the `pthread_mutex_t` corruption may have happened. Kodi's PVR code is broken in so many ways, it falls apart every time I look at it. A while ago, I had an idea: write a fake PVR plugin that hammers hard on Kodi, and tries to reveal all timing/locking/... bugs. Maybe it will be able to reproduce this one as well (we'll never know, because we cannot identify this bug exactly when it happens again). So this would be my approach. 

My mid- and long-term approach: document Kodi's thread-safety, lock dependencies, subsystem layers, review the exisitng code and fix all the policy violations.

**Pull request closed without merging**